File: codes/python/inheritance/multiple/multiple2.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

class Person(object):
    def __init__(self, name):
        self.name = name
        self.serial = 5

    def info(self):
        print("Person name: " + self.name)
        print("Person serial number: " + str(self.serial))

class Machine(object):
    def __init__(self, serial):
        self.serial = serial

    def info(self):
        print("Machine serial number: " + str(self.serial))

class Android(Machine, Person):
    def __init__(self, name, serial):
        Person.__init__(self, name)
        Machine.__init__(self, serial)

class Cyborg(Person, Machine):
    def __init__(self, name, serial):
        Machine.__init__(self, serial)
        Person.__init__(self, name)


p = Person("person1")
p.info()
print("=====================")
m = Machine(17)
m.info()
print("=====================")
a = Android("android1", 21)
a.info()
print("=====================")
c = Cyborg("cyborg1", 27)
c.info()
print("=====================")

File: codes/ruby/inheritance/multiple/multiple2.rb
#!/usr/bin/ruby -w

module PersonMixin
    def info; puts "my name is #{@name}"; end
end

module MachineMixin
    def info; puts "my serial number: #{@serial}"; end
end

class Person
    include PersonMixin
    def initialize(name)
        @name = name
    end
end

class Machine
    include MachineMixin
    def initialize(serial)
        @serial = serial
    end
end

class Android < Machine
    alias :parentinfo :info
    include PersonMixin
    alias :info2 :info
    def initialize(name, serial)
        super(serial)
        @name = name
    end
    def info
        parentinfo
        info2
    end
end

class Cyborg < Person
    alias :parentinfo :info
    include MachineMixin
    alias :info2 :info
    def initialize(name, serial)
        super(name)
        @serial = serial
    end
    def info
        parentinfo
        info2
    end
end

# ====== main =======
# ===================

p = Person.new("person1")
p.info
puts "====================="
m = Machine.new(21)
m.info
puts "====================="
a = Android.new("android1", 28)
a.info
puts "====================="
c = Cyborg.new("cyborg1", 32)
c.info
puts "====================="
